### DISCUSSION TITLE
Unable to update different renderers in a loop

### DISCUSSION THREAD

**Discussion by jkissi** [marcomusy/vedo#522] [Q&A] [answered]
Hi

Having trouble updating the renderer with new pointdata in a loop: 

![multicurve_edit](https://user-images.githubusercontent.com/8246894/140839002-90c6b654-5be9-48c2-877a-fc69fe464f96.png)

All these images, should be displaying different data but, as you can see, they are showing the same data. This is what I have so far:

```
    lst_df= ['pcx', 'pcy', 'pcz']
    lst_curve= ['XCurvature', 'YCurvature', 'ZCurvature']
    lst_str= ['XCurvature', 'YCurvature', 'ZCurvature']


    # initialise the plot object
    vp = ve.Plotter(bg='black', shape=[1, len(lst_curve)], sharecam=1)

    # grab the feature space
    df = struct.fuzzy['vars']['cloud_df']

    # grab the xyz
    points = fsp.select_feature_space(dframe=df, coltitles=['zx', 'zy', 'zz']).to_numpy()

    # grab the nxnynz
    normals = fsp.select_feature_space(dframe=df, coltitles=['nx','ny','nz']).to_numpy()

    # generate to the Points obj
    vpts = ve.Points(points,  r=3)

    # generate normal lines
    norms = ve.Lines(vpts, normals, alpha=0.25, scale=0.02)

    # for the number of curvature quantities listed
    for i in np.arange(0, len(lst_curve)):

        # grab the curvature series
        curv = fsp.select_feature_space(dframe=df, coltitles=[lst_df[i]]).to_numpy()

        # add to the pointdata dict
        vpts.pointdata[lst_curve[i]] = curv

        # select the dictionary for display
        vpts.pointdata.select(lst_curve[i])

        # generate colourbar
        vpts.cmap('jet', lst_curve[i]).addScalarBar()

        # show
        if i != len(lst_curve) - 1:
            # render current pointdata dict
            vp.render()
            vp.show(vpts, norms, lst_str[i], at=int(i), axes=1)
        else:
            # render current pointdata dict
            vp.render()
            vp.show(vpts, norms, lst_str[i], at=int(i) if len(lst_curve) > 1 else None, interactive=True, axes=1).close()
```
So the idea is that I have three different curvatures and I want each rendered in a seperate subplot. I assumed this is how I would do that in a loop, but I always seem to get the same plot rendered three times. Any ideas where I going wrong?

Cheers
J

**Comment by marcomusy** [comment]
the problem is that you always use the same pointcloud! So the first 2 renderers are also modified because you modify the **same** object.
To fix this you can use `vpts.clone()` (or try `vpts.clone(deep=False)`), to have a working copy of the same cloud.

**Reply by jkissi** [reply]
Works like magic!!
![multicurve_fixed_edit](https://user-images.githubusercontent.com/8246894/140857873-6333da35-4914-4bbe-b9bf-e5170e6b73c1.png)

Thanks again!
Have a quick question: am thinking of adding a button the plot to switch views between my classification and curvature map. I figure the button function would flip between different pointdata dictionaries and render, as both my class data and curvature data are in pointdata fields. I take it that I can't have these be in the same object? I would have to create a clone and have the button switch between the two?

J

**Comment by marcomusy** [comment, marked as answer]
nice!

yes - you can add a `Button` which calls a `plt.remove(old_pcl).add(new_pcl)`

`vedo --search button`
